File: apps/web/src/extraction/mensa.ts
const BASE_URL = "https://tum-dev.github.io/eat-api/en";

interface DayMenu {
  date: string;
  [key: string]: unknown;
}

interface WeekMenu {
  days?: DayMenu[];
  [key: string]: unknown;
}

export async function getAllCanteens(): Promise<unknown> {
  const res = await fetch(`${BASE_URL}/enums/canteens.json`);
  if (!res.ok) throw new Error(`Failed to fetch canteens: ${res.status}`);
  return res.json();
}

export async function getMenuWeek(canteenId: string, year: number, week: number): Promise<WeekMenu> {
  const weekStr = String(week).padStart(2, "0");
  const res = await fetch(`${BASE_URL}/${canteenId}/${year}/${weekStr}.json`);
  if (!res.ok) throw new Error(`Failed to fetch menu for week ${week}: ${res.status}`);
  return res.json();
}

export async function getMenuDay(canteenId: string, targetDate: Date): Promise<DayMenu | null> {
  const year = targetDate.getFullYear();
  const week = getISOWeek(targetDate);
  const menuWeek = await getMenuWeek(canteenId, year, week);

  const targetStr = formatDate(targetDate);
  return menuWeek.days?.find((d) => d.date === targetStr) ?? null;
}

function getISOWeek(date: Date): number {
  const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
  const dayNum = d.getUTCDay() || 7;
  d.setUTCDate(d.getUTCDate() + 4 - dayNum);
  const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  return Math.ceil(((d.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
}

function formatDate(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, "0");
  const d = String(date.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

File: apps/web/src/extraction/mensa.py
import requests
from datetime import date, datetime

BASE_URL = "https://tum-dev.github.io/eat-api/en"

def get_all_canteens():
    """Returns a list of all available canteens."""
    url = f"{BASE_URL}/enums/canteens.json"
    response = requests.get(url)
    response.raise_for_status()
    return response.json()

def get_menu_week(canteen_id: str, year: int, week: int):
    """Returns the menu of a canteen for a specific week."""
    url = f"{BASE_URL}/{canteen_id}/{year}/{week:02d}.json"
    response = requests.get(url)
    response.raise_for_status()
    return response.json()

def get_menu_day(canteen_id: str, target_date: date):
    """Returns the menu of a canteen for a specific date.
    
    Args:
        canteen_id: ID of the canteen
        target_date: Date to get menu for (date object)
    
    Returns:
        Dictionary with the day's menu, or None if not found
    """
    year, week, _ = target_date.isocalendar()
    menu_week = get_menu_week(canteen_id, year, week)
    
    target_str = target_date.strftime("%Y-%m-%d")
    
    for day_data in menu_week.get("days", []):
        if day_data.get("date") == target_str:
            return day_data
    return None


